### ISSUE TITLE
Loss doesn't decrease when using Adasum with model parallelism

### ISSUE THREAD

**Issue by talkhaldi** [horovod/horovod#2244]
**Environment:**
1. Framework: PyTorch
2. Framework version: 1.6.0
3. Horovod version: 0.20.0
4. MPI version: 4.0.3
5. CUDA version: 10.1.243
6. NCCL version: 2.7.8-1
7. Python version: 3.6.5
8. Spark / PySpark version: -
9. OS and version: Ubuntu 16.04.6
10. GCC version: 5.5.0

**Bug report:**

The model I'm using is too large to fit in a single GPU (joint 2 large transformers training), so I'm splitting the model to two GPUs on device1 and device2 as follows:

```
...
device_rank = args.local_rank

if args.shard_model:
     device_rank = args.local_rank * 2
torch.cuda.set_device(device_rank)
device = torch.device("cuda", device_rank)

if args.shard_model:
     device2 = torch.device("cuda", device_rank + 1)   
else:
     device2 = device

args.device = device
args.device2 = device2
....
```
With this I confirmed that, having 8 GPUs, every process (mpirun -np 4) runs on even GPUs, and they load half of the model on the next GPU (odd numbered).

```
class BigModel(torch.nn.Module):
   def __init__(self, args....):
      super().__init__()
      self.device = args.device
      self.device2 = args.device2
       
      self.transformer1 =  model_class.from_pretrained(....)

      self.transformer2 =  model_class.from_pretrained(....)
   
  def shard_model(self):
      self.transformer2.to(self.device2)

  def forward(self, inputs...):
      outs = self.transformer1(inputs)
      
      loss1, logits1 = outs

      inputs2 = logits1_to_input2()
      inputs2.to(self.device2)
     
      outs2 = self.transformer2(inputs2)
      loss2 = outs2

      loss1 = loss1.to(self.device2)

      return   loss1 + loss2
```

I load the model to devices like:

```
model = BigModel(args)

model.to(args.device)
model.shard_model()
```

I set up training following Horovod instructions.. (hvd.DistributedOptimizer, lr * hvd.local_size() when using adasum... etc)

**Expected behavior:**

Loss decreases normally towards convergence.

**What's actually happening:**

Loss starts at around 8 and is stuck around 8 (waited for an hour and a half)

Now to rule out several things, the following settings decrease loss normally:

- Base version of transformers, single GPU (no distributed or hvd) (Note: Large version is what can't fit into GPU)
- Base, torch.distributed package
- Base, mpirun with Horovod
- Base, Sharded to two GPUs as the code above (no distributed or hvd)
- Base, Sharded to two GPUs + Horovod, No adasum

Then when I try  Base, Sharded with Horovod, with adasum, the problem happens.

Any suggestion on why is this happening?

I appreciate the help!

**Comment by lengoanhcat** [comment]
Environment:

Framework: Tensorlfow
Framework version: 1.15.0
Horovod version: 0.20.0
MPI version: 4.0.3
CUDA version: 10.1.243
NCCL version: 2.7.8-1
Python version: 3.7.7
Spark / PySpark version: -
OS and version: Ubuntu 18.04
GCC version: 7.5.0

I install horovod with the following command.

```
HOROVOD_GPU_OPERATIONS=NCCL pip install --no-cache-dir horovod
```

I just simply run a provided example with / wihtout adasum and save operations into time-lines

``` 
horovodrun -np 2 --timeline-filename ~/tensorflow_mnist_adasum.json python tensorflow_mnist.py --use-adasum
horovodrun -np 2 --timeline-filename ~/tensorflow_mnist_average.json python tensorflow_mnist.py
```

When comparing two time-lines and its processes ( photo and json files of timelines attached ), an optimizer with use-adasum option does not call allreduce at all. 

![tensorflow_mnist_timeline](https://user-images.githubusercontent.com/1382227/92549589-f5bf5b00-f29c-11ea-8f18-62701e0dfe7e.PNG)

[tensorflow_mnist_average.txt](https://github.com/horovod/horovod/files/5192376/tensorflow_mnist_average.txt)
[tensorflow_mnist_adasum.txt](https://github.com/horovod/horovod/files/5192377/tensorflow_mnist_adasum.txt)

Is it a bug ? or Have I done something wrong here ?

Regards,

Cat Le
